File: projections/blazor/ui/hlp.ui.scheduler/SchedulerTypes.cs
namespace Harborline.UIAdapters.Blazor.Components.Scheduling;

public enum SchedulerViewType { Day, Week, Month, Agenda }
public enum SchedulerWorkDay { Sunday, Monday, Tuesday, Wednesday, Thursday, Friday, Saturday }
public sealed record SchedulerView(SchedulerViewType Type, string? Title = null);
public sealed record SchedulerWorkingHours(int Start, int End);

public sealed record SchedulerEvent(
    object Id,
    string Title,
    DateTimeOffset Start,
    DateTimeOffset End,
    bool AllDay = false,
    string? Color = null,
    string? Description = null,
    object? Resource = null,
    string? RecurrenceRule = null,
    object? RecurrenceId = null,
    IReadOnlyList<DateTimeOffset>? RecurrenceExceptions = null,
    DateTimeOffset? OriginalStart = null,
    IReadOnlyDictionary<string, object?>? Fields = null);

public sealed record SchedulerModelFields(
    string Id = "id",
    string Title = "title",
    string Start = "start",
    string End = "end",
    string AllDay = "allDay",
    string Description = "description",
    string Color = "color",
    string RecurrenceRule = "recurrenceRule",
    string RecurrenceId = "recurrenceId",
    string RecurrenceExceptions = "recurrenceExceptions",
    string OriginalStart = "originalStart");

public static class SchedulerModel
{
    public static SchedulerEvent Normalize(IReadOnlyDictionary<string, object?> item, SchedulerModelFields? fields = null)
    {
        ArgumentNullException.ThrowIfNull(item);
        fields ??= new();
        object Required(string key) => item.TryGetValue(key, out var value) && value is not null ? value : throw new InvalidOperationException("invalid-model-field");
        T? Optional<T>(string key) => item.TryGetValue(key, out var value) && value is T typed ? typed : default;
        var id = Required(fields.Id);
        var title = Required(fields.Title)?.ToString() ?? throw new InvalidOperationException("invalid-model-field");
        var start = ToDate(Required(fields.Start));
        var end = ToDate(Required(fields.End));
        return new(id, title, start, end,
            Optional<bool>(fields.AllDay), Optional<string>(fields.Color), Optional<string>(fields.Description),
            item.TryGetValue("resource", out var resource) ? resource : null,
            Optional<string>(fields.RecurrenceRule), item.TryGetValue(fields.RecurrenceId, out var recurrenceId) ? recurrenceId : null,
            Optional<IReadOnlyList<DateTimeOffset>>(fields.RecurrenceExceptions), Optional<DateTimeOffset?>(fields.OriginalStart),
            new Dictionary<string, object?>(item));
    }

    private static DateTimeOffset ToDate(object value) => value switch
    {
        DateTimeOffset date => date,
        DateTime date => new(date),
        string text when DateTimeOffset.TryParse(text, out var date) => date,
        _ => throw new InvalidOperationException("invalid-model-field"),
    };
}

File: projections/react/ui/hlp.ui.scheduler/src/Scheduler.types.ts
import type * as React from 'react'

export type SchedulerViewType = 'day' | 'week' | 'month' | 'agenda'
export type SchedulerEventId = string | number

export interface SchedulerView {
  readonly type: SchedulerViewType
  readonly title?: string
}

export interface SchedulerEvent {
  readonly id: SchedulerEventId
  readonly title: string
  readonly start: Date
  readonly end: Date
  readonly allDay?: boolean
  readonly color?: string
  readonly description?: string
  readonly resource?: SchedulerEventId
  readonly recurrenceRule?: string
  readonly recurrenceId?: SchedulerEventId
  readonly recurrenceExceptions?: readonly Date[]
  readonly originalStart?: Date
  readonly [key: string]: unknown
}

export interface SchedulerModelFields {
  readonly id?: string
  readonly title?: string
  readonly start?: string
  readonly end?: string
  readonly allDay?: string
  readonly description?: string
  readonly color?: string
  readonly resource?: string
  readonly recurrenceRule?: string
  readonly recurrenceId?: string
  readonly recurrenceExceptions?: string
  readonly originalStart?: string
}

export interface SchedulerWorkingHours {
  readonly start: number
  readonly end: number
}

/** Date#getDay convention: Sunday is 0 and Saturday is 6. */
export type SchedulerWorkDay = 0 | 1 | 2 | 3 | 4 | 5 | 6

export interface SchedulerProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'children' | 'onChange'> {
  readonly data: readonly SchedulerEvent[] | readonly Record<string, unknown>[]
  readonly view?: SchedulerViewType
  readonly defaultView?: SchedulerViewType
  readonly onViewChange?: (view: SchedulerViewType) => void
  readonly views?: readonly SchedulerView[]
  readonly onEventAdd?: (event: SchedulerEvent) => void
  readonly onEventUpdate?: (event: SchedulerEvent) => void
  readonly onEventDelete?: (event: SchedulerEvent) => void
  readonly onDateChange?: (date: Date) => void
  readonly selectedDate?: Date
  readonly defaultDate?: Date
  readonly now?: Date
  readonly workingHours?: SchedulerWorkingHours
  readonly workDayStart?: number
  readonly workDayEnd?: number
  readonly workDays?: readonly SchedulerWorkDay[]
  readonly readOnly?: boolean
  readonly modelFields?: SchedulerModelFields
  readonly accessibleName?: string
}

export interface ParsedSchedulerRRule {
  readonly freq: 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY'
  readonly interval: number
  readonly count?: number
  readonly until?: Date
  readonly byDay?: readonly { readonly ordinal?: number; readonly day: SchedulerWorkDay }[]
  readonly byMonthDay?: readonly number[]
  readonly byMonth?: readonly number[]
}
